### PR TITLE
Slider fix

### PR DESCRIPTION
**Pull request by DanielPerssonProos**
New slider implementation. Fixed a few scaling related bugs and wrote some tests on other methods as well.

### CODE REVIEW

File: ViAn/GUI/mainwindow.cpp [hunk 1]
```diff
@@ -217,28 +235,62 @@ void MainWindow::resizeEvent(QResizeEvent* event) {
...
 /**
+ * @brief MainWindow::on_slider_moving
+ * Sends frame numbers to the video player whilst the slider is moving
+ */
+void MainWindow::on_slider_moving(){
+    QPoint local_mouse_pos = ui->videoSlider->mapFromGlobal(QCursor::pos());
+    float pos_ratio = local_mouse_pos.x() / (float )ui->videoSlider->size().width();
+    int slider_range = ui->videoSlider->maximum() - ui->videoSlider->minimum();
+    int slider_pos_under_mouse = ui->videoSlider->minimum() + slider_range * pos_ratio;
+
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: This code is very similar to code in `on_slider_click`. Maybe you could extract it t oa separate method, or call `on_slider_click` from `on_slider_moving`.

File: ViAn/Video/video_player.cpp [hunk 2]
```diff
@@ -4,7 +4,9 @@
 #include "GUI/mainwindow.h"
 #include <thread>
 #include <QWaitCondition>
+#include <QDesktopWidget>
 #include <qpainter.h>
+#include <qdebug.h>
```
Comment (parso619) [inline ViAn/Video/video_player.cpp]: Is `#include <qdebug.h>` needed? (I tried commenting it out and it was no prolem.)

File: ViAn/GUI/mainwindow.cpp [hunk 3]
```diff
@@ -2,13 +2,15 @@
...
 #include <iostream>
+#include <sstream>
 #include <QCloseEvent>
 #include <QColorDialog>
 #include <chrono>
 #include <thread>
 #include "icononbuttonhandler.h"
 #include "inputwindow.h"
 #include "Video/shapes/shape.h"
+#include <qdebug.h>
```
Comment (parso619) [inline ViAn/GUI/mainwindow.cpp]: Is `#include <qdebug.h>` needed? (I tried commenting it out and it was no prolem.)

File: ViAn/Video/video_player.cpp [hunk 4]
```diff
@@ -125,23 +143,32 @@ void video_player::convert_frame(bool scale) {
...
         processed_frame = analysis_area->draw(processed_frame);
     }
-    processed_frame = zoom_frame(processed_frame);
-    processed_frame = contrast_frame(processed_frame);
 
-    if (scale && (frame_width != capture.get(CV_CAP_PROP_FRAME_WIDTH) || frame_height != capture.get(CV_CAP_PROP_FRAME_HEIGHT))) {
-        processed_frame = scale_frame(processed_frame);
-    }
+    zoom_frame(processed_frame).copyTo(processed_frame);
+    contrast_frame(processed_frame).copyTo(processed_frame);
```
Comment (parso619) [inline ViAn/Video/video_player.cpp]: Why have you changed to `copyTo` here, isn't that less effective?

**Comment by DanielPerssonProos** [comment]
In answer to parso619s comment about `set_current_frame_num`: 
The method `set_current_frame_num` is used many times internally in video_player class and is the method that actually fetches a new frame. `on_set_playback_frame` is a slot and is intended to be called via a signal from other threads. `on_set_playback_frame` also uses `set_current_frame_num` indirectly via the `update_frame` method. That `set_current_frame_num` is called from MainWindow is not correct and I will replace that call with the correct one. I will also make `set_current_frame_num` private in video_player.

File: ViAn/Video/video_player.h [hunk 5]
```diff
@@ -95,31 +96,37 @@ public slots:
...
+    int screen_height;
 
     double frame_rate;
     double speed_multiplier = DEFAULT_SPEED_MULT;
 
     bool video_stopped = false;
     bool video_paused;
     bool choosing_zoom_area = false;
+    bool set_new_frame = false;
+    bool slider_moving = false;
```
Comment (DanielPerssonProos) [inline ViAn/Video/video_player.h]: Removed it

File: ViAn/Test/test_video_player.cpp [hunk 6]
```diff
@@ -331,3 +324,97 @@ void test_video_player::test_set_stop_video() {
...
+    mvideo->video_paused = true;
+    QVERIFY(mvideo->set_current_frame_num(LEGAL_FRAME));
+    QVERIFY(mvideo->capture.get(CV_CAP_PROP_POS_FRAMES) == LEGAL_FRAME+1);
+    QVERIFY(mvideo->frame.elemSize()*mvideo->frame.cols*mvideo->frame.rows != 0);
+
+    mvideo->video_paused = false;
+    mvideo->new_frame_num = -1;
+    mvideo->set_new_frame = false;
+    mvideo->set_current_frame_num(LEGAL_FRAME);
+    QVERIFY(mvideo->new_frame_num == LEGAL_FRAME && mvideo->set_new_frame == true);
```
Comment (DanielPerssonProos) [inline ViAn/Test/test_video_player.cpp]: Removed it

File: ViAn/Video/video_player.cpp (hunk 2 above)
Comment (DanielPerssonProos) [inline ViAn/Video/video_player.cpp]: Removed it

File: ViAn/Video/video_player.cpp (hunk 4 above)
Comment (DanielPerssonProos) [inline ViAn/Video/video_player.cpp]: I changed it primarily for stability. I looked into it further an found that there were no improvements in stability when using `copyTo` so I changed it back.

File: ViAn/GUI/mainwindow.cpp (hunk 3 above)
Comment (DanielPerssonProos) [inline ViAn/GUI/mainwindow.cpp]: Removed it

**Comment by parso619** [comment]
Nice!

**Pull request merged**